pyop2/codegen/rep2loopy.py

```diff
@@ -85,6 +91,210 @@ def petsc_function_lookup(target, identifier):
+                arg_id_to_dtype={-1: NumpyType(mat_dtype, target=loopy.CTarget()), 0: NumpyType(mat_dtype, target=loopy.CTarget()), 1: NumpyType(mat_dtype, target=loopy.CTarget()), 2: NumpyType(numpy.int32, target=loopy.CTarget())}),
+                callables_table)
+
+    def emit_call_insn(self, insn, target, expression_to_code_mapper):
+        assert self.is_ready_for_codegen()
+        assert isinstance(insn, loopy.CallInstruction)
+
+        parameters = insn.expression.parameters
+
+        if type(parameters) != list:
+            parameters = list(parameters)
+        par_dtypes = [self.arg_id_to_dtype[i] for i, _ in enumerate(parameters)]
+
+        parameters.append(insn.assignees[0])
+        par_dtypes.append(self.arg_id_to_dtype[1])
+        mat_descr_A = self.arg_id_to_descr[0]
+
+        arg_c_parameters = [
+            expression_to_code_mapper(
+                par,
+                PREC_NONE,
+                dtype_to_type_context(target, par_dtype),
+                par_dtype
+            ).expr
+            if isinstance(par, SubArrayRef) else
+            expression_to_code_mapper(
+                par,
+                PREC_NONE,
+                dtype_to_type_context(target, par_dtype),
+                par_dtype
+            ).expr
+            for par, par_dtype in zip(parameters, par_dtypes)
+        ]
+        c_parameters = []
+        c_parameters.insert(0, arg_c_parameters[-1])  # out
+        c_parameters.insert(1, arg_c_parameters[0])  # A
+        c_parameters.insert(2, arg_c_parameters[1])  # B
+        c_parameters.insert(3, mat_descr_A.shape[1])  # n
+        return var(self.name_in_target)(*c_parameters), False
+
+    def generate_preambles(self, target):
+        assert isinstance(target, CTarget)
+        code = """#include <string.h>
+            #include <stdio.h>
+            #include <stdlib.h>
+            #ifndef Solve_HPP
+            #define Solve_HPP
+            #define BUF_SIZE 30
+
+            static PetscBLASInt ipiv_buffer[BUF_SIZE];
+            static void solve(PetscScalar* restrict out, const PetscScalar* restrict A, PetscScalar* restrict B, PetscBLASInt N)
+            {
+                PetscBLASInt info;
+                PetscBLASInt *ipiv = N <= BUF_SIZE ? ipiv_buffer : malloc(N*sizeof(*ipiv));
+                memcpy(out,B,N*sizeof(PetscScalar));
+                PetscBLASInt NRHS;
+                NRHS=1;
+                LAPACKgesv_(&N,&NRHS,A,&N,ipiv,out,&N,&info);
+                if(info!=0)
+                    fprintf(stderr,\"Gesv throws nonzero info.\");
+                if ( N > BUF_SIZE ) {
+                    free(ipiv);
+                }
+            }
+            #endif
+        """
+
+        yield("solve", "#include <petscsys.h>\n#include <petscblaslapack.h>\n"+code)
+        return
+
+
+def sol_fn_lookup(target, identifier):
+    if identifier == 'solve':
+        return SolveCallable(name='solve')
+    else:
+        return None
```

Comment:
These two callables (they are quite related) share a lot of code. Can we refactor so that they actually do so, and just subclass the bits that are different?